// interrogator/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
let inquirerCache;
async function inquirer() {
    return await (inquirerCache ||
        (inquirerCache = import('inquirer').then((i) => i.default)));
}
// end inquirer import
async function typedPrompt(question) {
    const { prompt } = await inquirer();
    return (await prompt(question)).value;
}
async function replaceSeparators(choices) {
    const i = await inquirer();
    return choices.map((c) => c instanceof Separator ? new i.Separator(c.line) : c);
}
class Separator {
    constructor(line) {
        this.type = 'separator';
        this.line = line;
    }
}
function getDefaultValue(choices, defaultValue) {
    if (defaultValue === undefined)
        return undefined;
    if (choices.every((c) => typeof c === 'string'))
        return defaultValue;
    const result = choices.findIndex((c) => c.value === defaultValue);
    if (result === -1)
        return undefined;
    return result;
}
async function list(message, choices, defaultValue) {
    return await typedPrompt({
        type: 'list',
        name: 'value',
        message,
        choices: await replaceSeparators(choices),
        default: getDefaultValue(choices, defaultValue),
    });
}
async function rawList(message, choices, defaultValue) {
    return await typedPrompt({
        type: 'rawlist',
        name: 'value',
        message,
        choices: await replaceSeparators(choices),
        default: getDefaultValue(choices, defaultValue),
    });
}
async function expand(message, choices, defaultValue) {
    return await typedPrompt({
        type: 'expand',
        name: 'value',
        message,
        choices: await replaceSeparators(choices),
        default: getDefaultValue(choices, defaultValue),
    });
}
async function checkboxes(message, choices) {
    return await typedPrompt({
        type: 'checkbox',
        name: 'value',
        message,
        choices: await replaceSeparators(choices),
    });
}
async function confirm(message, defaultValue = true) {
    return await typedPrompt({
        type: 'confirm',
        name: 'value',
        message,
        default: defaultValue,
    });
}
async function input(message, defaultValue, validate) {
    return await typedPrompt({
        type: 'input',
        name: 'value',
        message,
        default: defaultValue,
        validate,
    });
}
async function password(message, defaultValue, validate) {
    return await typedPrompt({
        type: 'password',
        name: 'value',
        message,
        default: defaultValue,
        validate,
    });
}
async function editor(message, defaultValue = '') {
    return await typedPrompt({
        type: 'editor',
        name: 'value',
        message,
        default: defaultValue,
    });
}
exports.Separator = Separator;
exports.list = list;
exports.rawList = rawList;
exports.expand = expand;
exports.checkboxes = checkboxes;
exports.confirm = confirm;
exports.input = input;
exports.password = password;
exports.editor = editor;
