// @databases/pg-schema-introspect/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.sql = exports.connect = exports.default = exports.TypeKind = exports.TypeCategory = exports.ForeignKeyMatchType = exports.ForeignKeyAction = exports.ConstraintType = exports.ClassKind = void 0;

var ClassKind_1 = require("./enums/ClassKind");

Object.defineProperty(exports, "ClassKind", {
  enumerable: true,
  get: function () {
    return ClassKind_1.default;
  }
});

var ConstraintType_1 = require("./enums/ConstraintType");

Object.defineProperty(exports, "ConstraintType", {
  enumerable: true,
  get: function () {
    return ConstraintType_1.default;
  }
});

var ForeginKeyAction_1 = require("./enums/ForeginKeyAction");

Object.defineProperty(exports, "ForeignKeyAction", {
  enumerable: true,
  get: function () {
    return ForeginKeyAction_1.default;
  }
});

var ForeignKeyMatchType_1 = require("./enums/ForeignKeyMatchType");

Object.defineProperty(exports, "ForeignKeyMatchType", {
  enumerable: true,
  get: function () {
    return ForeignKeyMatchType_1.default;
  }
});

var TypeCategory_1 = require("./enums/TypeCategory");

Object.defineProperty(exports, "TypeCategory", {
  enumerable: true,
  get: function () {
    return TypeCategory_1.default;
  }
});

var TypeKind_1 = require("./enums/TypeKind");

Object.defineProperty(exports, "TypeKind", {
  enumerable: true,
  get: function () {
    return TypeKind_1.default;
  }
});

var getSchema_1 = require("./getSchema");

Object.defineProperty(exports, "default", {
  enumerable: true,
  get: function () {
    return getSchema_1.default;
  }
});
Object.defineProperty(exports, "connect", {
  enumerable: true,
  get: function () {
    return getSchema_1.connect;
  }
});
Object.defineProperty(exports, "sql", {
  enumerable: true,
  get: function () {
    return getSchema_1.sql;
  }
});
